### PR TITLE
Phase4

### PR DESCRIPTION
**Pull request by belucid**
https://trello.com/c/rjjcRKFY

This PR covers the next wave of commenting functionality, namely:

- Comment delete and update API support for the UI (no UI support yet) with HATEOAS link support for the same
- WebSocket events for comment update and delete (no UI support yet)
- Text in the initial in channel Slack mirrored comment that tells the Slack channel what's going on, with a link to the OC entry
- Returning comments in created-at order in API response
- Store bot-token in `slack-thread` on entry for later Slack->OC mirror authorship lookup
- Attributing authorship when mirroring comments from Slack->OC with a cache

Supporting lib PR: https://github.com/open-company/open-company-lib/pull/14

Supporting storage PR: https://github.com/open-company/open-company-api/pull/85

Supporting master commits:
 https://github.com/belucid/open-company-infrastructure/commit/aee294104525a80399f5e5913f9e9f78bf40bdf8
https://github.com/belucid/open-company-infrastructure/commit/e68b94cf2a98ea08d1b90d0316bdb69b4301d0eb

To test:

- Use cURL or Paw or what have you to update a comment you create in the UI (e.g. PATCH `http://localhost:3002/orgs/9f75-44e3-adb9/boards/5149-4136-b2af/topics/update/entries/de90-46bf-a746/comments/0f1e-41dd-8c92`). Is the body of the comment updated? How about the updated-at?
- Watch the WebSocket frames for the above in the web client, do you see the WebSocket event for the comment update?
- Use cURL or Paw or what have you to delete a comment you create in the UI (e.g. DELETE `http://localhost:3002/orgs/9f75-44e3-adb9/boards/5149-4136-b2af/topics/update/entries/de90-46bf-a746/comments/0f1e-41dd-8c92`). Is the comment deleted?
- Watch the WebSocket frames for the above in the web client, do you see the WebSocket event for the comment delete?
- Create a new comment in OC on a new entry, is it mirrored to Slack? Is it mirrored with a statement about what's happening and a link to the OC entry? (OC entry link won't work yet as client doesn't have URL addressability for entries). Try it as a Slack OC user and an email OC user.
- Reply to that comment in the created Slack thread. Is the comment echoed back to OC (make sure the Slack webhook is in enabled state and ngrok is running!) and is the author of the Slack comment properly attributed w/ a name and avatar (e.g. not "Slack user").
- Reply again in the same thread and watch the interaction service output, did you notice a cache hit instead of a cache miss and no Slack lookup?
- Good... review the codes
- Merge
- Deploy
- Adopt a pet ferret from a shelter

### CODE REVIEW

File: src/oc/interaction/api/comments.clj [hunk 1]
```diff
@@ -26,8 +32,96 @@
...
+    
+    true)) ; no existing comment, so this will fail existence check later
+
+(defn author-match?
+  "Determine if the comment specified by the UUID was authored by the current user."
+  [conn ctx comment-uuid]
+  (if-let* [user-id (-> ctx :user :user-id)
+            existing-comment (or (:existing-comment ctx) (interact-res/get-interaction conn comment-uuid))
+            author-id (-> existing-comment :author :user-id)]
+    (if (= user-id author-id)
```
Comment (bago2k4) [inline src/oc/interaction/api/comments.clj:51]: Only the user that posted the comment can edit it at the moment right? an admin or author can't edit or delete it?

File: src/oc/interaction/api/comments.clj [hunk 2]
```diff
@@ -83,10 +177,19 @@
...
         [org-uuid board-uuid topic-slug entry-uuid]
-        (pool/with-pool [conn db-pool] (comment-list conn org-uuid board-uuid topic-slug entry-uuid))))))
\ No newline at end of file
+        (pool/with-pool [conn db-pool] (comment-list conn org-uuid board-uuid topic-slug entry-uuid)))
+
+      ;; Comment editing and removal
+      (ANY "/orgs/:org-uuid/boards/:board-uuid/topics/:topic-slug/entries/:entry-uuid/comments/:comment-uuid"
+        [org-uuid board-uuid topic-slug entry-uuid comment-uuid]
+        (pool/with-pool [conn db-pool] (comment-item conn org-uuid board-uuid topic-slug entry-uuid comment-uuid)))
+      (ANY "/orgs/:org-uuid/boards/:board-uuid/topics/:topic-slug/entries/:entry-uuid/comments/:comment-uuid/"
```
Comment (bago2k4) [inline src/oc/interaction/api/comments.clj:193]: Why 2 times the same route if it's accepting ANY method?

Comment (bago2k4) [reply src/oc/interaction/api/comments.clj:193]: also they looks exactly the same..

File: src/oc/interaction/async/slack_mirror.clj [hunk 3]
```diff
@@ -11,25 +11,105 @@
...
+
+(def echo-intro "commenting on:")
+(def proxy-intro "There's a comment on:")
+
+;; ----- Slack defaults -----
+
+(def default-slack-author {
+  :user-id "0000-0000-0000"
+  :name "Slack User"
+  :avatar-url nil})
```
Comment (bago2k4) [inline src/oc/interaction/async/slack_mirror.clj]: We can default the logo to one of our shape.. from the staging CDN for now:
https://do64qbk3yqbc.cloudfront.net/img/ML/user_avatar_red.png
and to the beta one when we will get there.
https://d1wc0stj82keig.cloudfront.net/img/ML/user_avatar_red.png
You can change the color with one of: blue, yellow, purple, green and red.
Also the CDN url get come from the config ENV vars, infra already has them.

File: src/oc/interaction/async/slack_mirror.clj [hunk 4]
```diff
@@ -11,25 +11,105 @@
...
+                          (:image_72 user-data)
+                          (:image_48 user-data)
+                          (:image_32 user-data)
+                          (:image_24 user-data)
+                          (get-in user-data [:profile :image_512])
+                          (get-in user-data [:profile :image_192])
+                          (get-in user-data [:profile :image_72])
+                          (get-in user-data [:profile :image_48])
+                          (get-in user-data [:profile :image_32])
+                          (get-in user-data [:profile :image_24])))]
```
Comment (bago2k4) [inline src/oc/interaction/async/slack_mirror.clj]: and default to the user avatar here like you do for the user-name if you don't get one from Slack.

**Review by bago2k4 — CHANGES_REQUESTED** [review]

File: src/oc/interaction/api/comments.clj (hunk 1 above)
Comment (belucid) [reply src/oc/interaction/api/comments.clj:51]: Right. Not sure if we'll ever let someone else edit or delete it, but we won't for now unless/until a customer brings it up and wants it.

File: src/oc/interaction/api/comments.clj (hunk 2 above)
Comment (belucid) [reply src/oc/interaction/api/comments.clj:193]: Only difference is the trailing `/`. We have the duplicated routes, with and w/o trailing `/` in most of the HTTP services. Seems to be a limitation of compojure that you can't create one route that accounts for both.

File: src/oc/interaction/async/slack_mirror.clj (hunk 3 above)
Comment (belucid) [reply src/oc/interaction/async/slack_mirror.clj]: Discussion in Slack.

File: src/oc/interaction/async/slack_mirror.clj (hunk 4 above)
Comment (belucid) [reply src/oc/interaction/async/slack_mirror.clj]: Discussion in Slack.

**Review by belucid — COMMENTED** [review]
Answers.

**Review by bago2k4 — APPROVED** [review]

**Pull request merged**